File: rpi_tph_monitor_sample/lcd/python3/lcdaqm.py
#!/usr/bin/env python3

"""
LCD AQM0802A-RN-GBW Control Module via I2C
 2017/10/22
"""

import smbus
import time

class LCDAQM:
    def __init__(self):
        self.i2c_addr = 0x3E
        self.i2c = smbus.SMBus(1)
        self.count = 0  # Count number of character in current line
        self.line = 1   # Current line number

    # I2C write one byte data to addr
    def write_address_onebyte(self, addr, data):
        data_list = [data]
        try:
            self.i2c.write_i2c_block_data(self.i2c_addr, addr, data_list)
        except IOError:
            return False

        time.sleep(50/1000000)

        return True

    # Initialize LCD
    def init_lcd(self):
        self.write_address_onebyte(0, 0x38)
        self.write_address_onebyte(0, 0x39)
        self.write_address_onebyte(0, 0x14)
        self.write_address_onebyte(0, 0x70)
        self.write_address_onebyte(0, 0x56)
        self.write_address_onebyte(0, 0x6C)
        time.sleep(0.25)
        self.write_address_onebyte(0, 0x38)
        self.write_address_onebyte(0, 0x0C)
        self.clear();

    # Print string on LCD
    def print(self, str):
        chars = list(str)
        for i in range(len(chars)):
            self.count += 1
            if self.count > 8:
                if self.line==2:
                    home()
                else:
                    sec_line()
            self.write_address_onebyte(0x40, ord(chars[i]))

    # Clear LCD and return to home
    def clear(self):
        self.count = 0
        self.line = 1
        self.write_address_onebyte(0, 0x1)
        time.sleep(0.002)

    # Return to home without deleting characters
    def home(self):
        self.count = 0
        self.line = 1
        self.write_address_onebyte(0, 0x2)
        time.sleep(0.002)

    # Go to second line
    def sec_line(self):
        self.count = 0
        self.line = 2
        self.write_address_onebyte(0, 0xC0)


def main():
    lcd = LCDAQM()
    lcd.init_lcd()

    lcd.print('AQM0802')
    time.sleep(1)
    lcd.clear()
    time.sleep(1)
    lcd.print('Test')
    time.sleep(1)
    lcd.sec_line()
    lcd.print('Success')
    time.sleep(1)

if __name__ == '__main__':
    main()

File: rpi_tph_monitor_sample/lcd/cpp/LCDAQM.cpp
//---------------------------------
// 2016/8/24

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>
#include <unistd.h>
#include <fcntl.h>
#include <wiringPi.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include "LCDAQM.h"


//---------------------------------
// Private Class Function

// Write control and data by I2C
void LCDAQM::writeData(uint8_t ctrl, uint8_t data){
	uint8_t buf[2];
	buf[0]=ctrl;
	buf[1]=data;
	write(i2c,buf,2);
	delayMicroseconds(50);
	return;
}



//---------------------------------
// Public Class Function

LCDAQM::LCDAQM(){
	count = 0;
	line = 1;
}


// Initialize LCD
//  Return 0 if success
uint8_t LCDAQM::init(){
	// Open I2C
	i2c = open("/dev/i2c-1",O_RDWR);
	if(i2c<0){
		return 1;
	}

	// Set I2C Address
	if(ioctl(i2c, I2C_SLAVE, i2cAdr)<0){
		close(i2c);
		return 1;
	}
	delay(50);

	writeData(0, 0x38);
	writeData(0, 0x39);
	writeData(0, 0x14);
	writeData(0, 0x70);
	writeData(0, 0x56);
	writeData(0, 0x6C);
	delay(250);
	writeData(0, 0x38);
	writeData(0, 0x0C);
	clear();

	return 0;
}

void LCDAQM::closeI2C(){
	close(i2c);
}

void LCDAQM::printStr(char*str){
	for(uint8_t i=0; i<16; i++){
		// Null character
		if(str[i]==0){
			return;
		}
		count++;
		if(count>8){
			if(line==2){
				home();
			}else{
				secLine();
			}
		}
		writeData(0x40, str[i]);
	}
}

void LCDAQM::clear(){
	count = 0;
	line = 1;
	writeData(0, 0x01);
	delay(2);
}


void LCDAQM::home(){
	count = 0;
	line = 1;
	writeData(0, 0x02);
	delay(2);
}


void LCDAQM::secLine(){
	count = 0;
	line = 2;
	writeData(0, 0xC0);
}









